src/SonarScanner.MSBuild.PreProcessor/JreCaching/JreResolver.cs

```diff
@@ -0,0 +1,116 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Threading.Tasks;
+using SonarScanner.MSBuild.Common;
+
+namespace SonarScanner.MSBuild.PreProcessor.JreCaching;
+
+// https://xtranet-sonarsource.atlassian.net/wiki/spaces/LANG/pages/3155001372/Scanner+Bootstrapping
+internal class JreResolver : IJreResolver
+{
+    private readonly ISonarWebServer server;
+    private readonly IJreCache cache;
+    private readonly ILogger logger;
+
+    public JreResolver(ISonarWebServer server, IJreCache cache, ILogger logger)
+    {
+        this.server = server;
+        this.cache = cache;
+        this.logger = logger;
+    }
+
+    public async Task<string> ResolveJrePath(ProcessedArgs args, string sonarUserHome)
+    {
+        logger.LogDebug(Resources.MSG_JreResolver_Starting);
+        if (!IsValid(args))
+        {
+            return null;
+        }
+
+        var metadata = await server.DownloadJreMetadataAsync(args.OperatingSystem, args.Architecture);
+        if (metadata is null)
+        {
+            logger.LogDebug(Resources.MSG_JreResolver_MetadataFailure);
+            return null;
+        }
+
+        var descriptor = metadata.ToDescriptor();
+        var isCachedResult = cache.IsJreCached(sonarUserHome, descriptor);
+        switch (isCachedResult)
+        {
+            case JreCacheHit hit:
+                logger.LogDebug(Resources.MSG_JreResolver_CacheHit, hit.JavaExe);
+                return hit.JavaExe;
+            case JreCacheMiss:
+                logger.LogDebug(Resources.MSG_JreResolver_CacheMiss);
+                return await DownloadJre(sonarUserHome, metadata, descriptor);
+            case JreCacheFailure:
+                logger.LogDebug(Resources.MSG_JreResolver_CacheFailure);
+                return null;
+        }
+
+        return null;
+    }
+
+    private async Task<string> DownloadJre(string sonarUserHome, JreMetadata metadata, JreDescriptor descriptor, bool retry = true)
+    {
+        var retrying = retry ? string.Empty : " Retrying...";
+        logger.LogDebug(Resources.MSG_JreResolver_DownloadAttempt, retrying);
+        var result = await cache.DownloadJreAsync(sonarUserHome, descriptor, () => server.DownloadJreAsync(metadata));
+        switch (result)
+        {
+            case JreCacheHit hit:
+                logger.LogDebug(Resources.MSG_JreResolver_DownloadSuccess, hit.JavaExe);
+                return hit.JavaExe;
+            default:
+                return retry
+                ? await DownloadJre(sonarUserHome, metadata, descriptor, false)
```

Comment:
@martin-strecker-sonarsource to answer your question on the other PR, here I am re-trying no matter the reason you failed.
If you want we can add more fine-grained control and re-try only on checksum validation problems, but IMO this is fine.